storage/innobase/lock/lock0lock.cc

```diff
@@ -3976,6 +3978,36 @@ dberr_t lock_table_for_trx(dict_table_t *table, trx_t *trx, lock_mode mode,
   return err;
 }
 
+/** Lock the child tables of a table.
+@param table    parent table
+@param trx      transaction
+@return error code */
+dberr_t lock_table_children(dict_table_t *table, trx_t *trx)
+{
+  dict_sys.freeze(SRW_LOCK_CALL);
+  std::vector<dict_table_t*> children;
+
+  for (auto f : table->referenced_set)
+    if (dict_table_t *child= f->foreign_table)
+    {
+      child->acquire();
+      children.emplace_back(child);
+    }
+  dict_sys.unfreeze();
+
+  dberr_t err= DB_SUCCESS;
+
+  for (auto child : children)
+    if ((err= lock_table_for_trx(child, trx, LOCK_X)) != DB_SUCCESS)
```

Comment:
Yet, another general comment about X lock for children. Is it really required ? We need to evaluate carefully.

1. Most cases, the presence of a referenced children would result in and error. For example normally for "DROP TABLE" ... 
ERROR 1451 (23000): Cannot delete or update a parent row: a foreign key constraint fails
1451: ER_ROW_IS_REFERENCED_2
So, there doesn't seem to be a good point to disrupt the operations on children first with possible waits on lock and finally returning and error. So, possibly may not need the locks in general case.

2. The special cases AFAICS are (from ha_innobase::delete_table code, we need to consider other DDL too)
* SET FOREIGN_KEY_CHECKS= OFF; It seems we try to drop the constraint from the child table. However, it is innodb only and the metadata in sql remains. So, seems half hearted and we might decide whether worth.
* DROP SCHEMA when both parent and child are in same schema. Even here I am not sure if we should really bother to lock the child.
* table->no_rollback() : Seems to be used for sequence with Innodb and FK should not be relevant.

Reply:
This entire function is a work-around for the lack of sufficient MDL acquisition on DDL operations in the SQL layer.